### PR TITLE
[Feature] Added the ability to add bootstrap tabs to forms

### PR DESCRIPTION
**Pull request by OwenMelbz**
This may or may-not be considered as backwards compatible, it modifies 3 template files:
- `form_content.blade.php` - this is the main modification, which includes markup for bootstraps tabs
- `create.blade.php` and `edit.blade.php` - this passes through the tabs to the `form_content.blade.php`

# Summary
This PR adds the ability for people to add basic organisational tabs.

# Preview
add_question____footy_finance_admin

# Usage

## Setting default layout
Globally within `config/backpack/crud.php`

```php
'tabs_horizontal' => true,
```

Entity level layout
```php
$this->crud->enableVerticalTabs();
$this->crud->disableVerticalTabs();
$this->crud->enableHorizontalTabs();
$this->crud->disableHorizontalTabs();
```

## Defining tabs
Simply call `$this->crud->addTab('TAB NAME');` to add tabs between fields, will contain all the proceeding fields, until either the end or another tab is created

```php
$this->crud->addTab('Basic');

$this->crud->addField([
    'label' => 'Question',
    'name' => 'question',
    'type' => 'text',
    'attributes' => ['required' => 'required']
]);

$this->crud->addTab('Option A');

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required']
]);

$this->crud->addField([
    'label' => 'Points',
    'name' => 'answer_1_value',
    'type' => 'number',
    'attributes' => ['required' => 'required']
]);

$this->crud->addField([
    'label' => 'Outcome',
    'name' => 'answer_1_reply',
    'type' => 'text',
    'attributes' => ['required' => 'required']
]);

$this->crud->addTab('Option B');

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_2',
    'type' => 'text',
    'attributes' => ['required' => 'required']
]);
```

## Removing tabs
```php
$this->crud->removeTab('Option A');
```

## Clearing tabs
```php
$this->crud->clearTabs();
```

### CODE REVIEW

**Comment by zschuessler** [comment]
Wow, great stuff. _Just_ implemented something similar this morning, but this is much nicer! Thanks for sharing!

**Comment by OwenMelbz** [comment]
@zschuessler hopefully it will prove useful! 

Personally I prefer vertical tabs typically as you can have more of them, but AdminLTE and the default spacing we have is a little different, but I'm sure people will find this enough for the time being!

Just wondering how it could be made anymore backwards compatible incase people have modified the 3 views. Maybe an "upgrade" guide would be needed - I guess technically it is backwards compatible, just users with modified views would not benefit from it until they update their templates.

Just have to see what @tabacitu thinks :) never know might get shipped with 3.2!

**Comment by tabacitu** [comment]
Just awesome. Great great work @OwenMelbz .

I've implemented a similar solution in a project, my syntax was a little bit different. Instead of:
```php
$this->crud->addTab('Option A');

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required']
]);
```

I had:
```php
$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Option 1',
]);
```

But my solution was miles away from yours, it was just a quick fix for that project. Not a complete feature, like yours. You might have your reasons for going this route, won't know until I dive in. Again - really really great job.

I'll take a hard look at this, especially how we can use it in addition to the multi_language feature. Because that would also involve tabs. But I think we can solve that pretty easily, by having the language tabs on top of the .box, and the custom tabs inside the .box, like you have them right now. I don't think expect that to be a problem.

I will give it a thorough look an get back to you. Right now, I see no reason why we wouldn't push this _before_ 3.2. It doesn't look like a breaking change. Plus, Laravel just changed its release cycle and it seems like Laravel 5.4 will launch in January.

Cheers!

**Comment by tabacitu** [comment]
Nope, you're totally right, people might have modified their 3 views, so it IS kind of a breaking change :-)

**Comment by OwenMelbz** [comment]
@tabacitu - I think it's a tricky one, as it won't break anything. It just won't do anything without tweaking your views.

If we did launch before 3.2 then maybe just a "FAQ" attached to it.

The reason I did it using just an 'addTab' was mainly inspired by the simplicity of ACF within Wordpress as it was dead simple and you didn't need to add an extra param to ever field. So thought it stuck to laravels principles of being more eloquent/fluent.

Would be interesting to see how others have implemented there's? As this was as simple as I could think of

**Comment by zschuessler** [comment]
Have now used this in several use cases, very very nice!

Architecture _may_ need to change to be more similar to @tabacitu's implementation. I have a requirement to not show a certain tab/fieldset on the update form _only_. The current API could take into account the `create/update/both` field option easy enough, but the entity CrudController would quickly become less readable. (From what I see anyway!)

**Comment by OwenMelbz** [comment]
@zschuessler cant you wrap the whole section around an if statement? so it wont even bother if its on the "create" form?

**Comment by zschuessler** [comment]
@OwenMelbz That's possible, but ignoring the `create/update/both` option on fields will leave a lot of devs scratching their heads (as I did!).

I rewrote the trait to function as more of a layout instead of a field. Eventually I'd like to see a layout management abstraction, since this feature quickly outgrows a php trait. I'm rewriting a feature for doing that, but a quickfix solution ended up with:

**Product Crud Controller**
```php
// Create tab navigation layout
$tabset = $this->crud->createTabNavigation([
  'tabs' => [
    [
      'label' => 'Content'
      // You can set tab pane/nav attributes here too
    ],
    [
      'label' => 'Pricing'
    ]
]);

/// Example field
$this->crud->addField(
  'name' => 'title',
  'tab'  => 'Content'
);
```

I also removed the need to set a 'crud' attribute in the form blade templates. In the parent crud controller class, you can set the tabs property for the view like so:

**Crud Controller - Parent Class**
```php
public function create()
{
  ...
  $this->data['tabs']   = $this->crud->getTabsetForView($this->data['fields']);
}
```

Then I added a layouts folder. The current simple view is the default layout, and then added a tabs layout blade template.

That gets me by for now. Interested in what others think about managing different types of layouts in the future. (Think of the different layouts AdminLTE offers - could be a nice feature)

**Comment by OwenMelbz** [comment]
@zschuessler I just realised I didn't actually understand what you meant on your second comment with the 'create/update/both'!

Now I understand what you mean, you're correct it was bugged.

I've actually updated this PR fixing, which may address your issue?

It also makes it more backwards compatible as I hadn't fully understood how those fields were populated in the first place but now I've dug around more I'd spotted!

So @tabacitu I think now its technically backwards compatible...You could release it without breaking anybodies sites I think.

But just like many of the others updates, they'd need to re-publish a file, in this case it would be the `form_content.blade.php` so that might be something to consider.

@zschuessler I wonder if you would be able to see if the latest addition does actually solve your original issue?

Thanks :)

**Comment by OwenMelbz** [comment]
pst @zschuessler ! did you get a chance to have a look/think? :D

**Comment by zschuessler** [comment]
Looks like it, good work!

**Comment by OwenMelbz** [comment]
Lovely :D hopefully it can make its way in at some point into a future release!

**Comment by imcodingideas** [comment]
I could use this, now :) Good PR.

**Comment by tabacitu** [comment]
Changed quite a few things. 

1) In the previous implementation, the fields were stored in 2 places: $create_fields / $update_fields AND $tabs; This would could have confused users and would definitely have become a nightmare in the long run. Now fields are stored in only one place ($create_fields / $update_fields) and may have an extra attribute, "tab".
2) We now get the tabs by parsing the fields and seeing which ones belong to which tab. If they don't belong to any tab, they're placed in the last tab.
3) The vertical/horizontal functionality was implemented as one object property, $horizontal, and by that we determined if it was vertical. Not cool :-) It would definitely have suffer changes in the future and we would create a breaking change. It's now called ```$tabsType```, and it returns ```horizontal``` or ```vertical```. Should leave plenty of room for the future.
4) New syntax/usage. In order to split your form into tabs, you just need to enable the functionality and choose a tab for each field.

```php
$this->crud->enableTabs();

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 1',
]); // creates tab called "Tab 1"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 2',
]); // creates tab called "Tab 2"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 1',
]); // adds it to "Tab 1"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 3',
]); // creates tab called "Tab 3"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    // no tab defined
]);  // adds it to "Tab 3"
```

Questions:
1) What do you think? Better? Easier to use/understand?
2) We could make this even easier to use, by auto-enabling the tabs functionality if a "tab" attribute is defined for a field. Should we? Does anyone see any downsides to this?

**Comment by MarcosBL** [comment]
This is AWESOME :D

**Comment by OwenMelbz** [comment]
@tabacitu my only thoughts would be the fact I liked the fluent syntax of `$this->crud->addTab()`; rather than passing it into each field :D

File: src/resources/views/form_content.blade.php [hunk 1]
```diff
@@ -42,13 +42,13 @@
             </style>
         @endpush
 
-        <div class="tab-container {{$horizontalTabs ? 'col-md-12' : 'col-md-3 m-t-10'}}">
+        <div class="tab-container {{ $horizontalTabs ? 'col-md-12' : 'col-md-3 m-t-10' }}">
 
             <div class="nav-tabs-custom">
-                <ul class="nav nav-tabs{{!$horizontalTabs ? ' nav-stacked' : ''}}" role="tablist">
-                    @foreach ($tabs as $k => $t)
+                <ul class="nav nav-tabs {{!$horizontalTabs ? ' nav-stacked' : ''}}" role="tablist">
```
Comment (OwenMelbz) [inline src/resources/views/form_content.blade.php]: pst, this adds double spacing into the classname :D

**Comment by tabacitu** [comment]
@OwenMelbz , yeah, I understand, but it's pretty atypical and I think it would cause more harm than good. I would expect a fluent Laravel syntax to work more like this:
```diff
- $this->crud->addTab('Tab 3');

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
-]);
+])->toTab('Tab 3');
```

I've never seen a Laravel method creating changes for the next Laravel methods. Except in blade files. I might be wrong, of course.

Plus, when they do, they're usually followed by an "end block" statement. Which you just KNOW people are going to forget :-)
```
@block
 // seen this, but only in Blade files 
@endblock

$this->something();
  // haven't really seen this
$this->endSomething();
```

So I really think specifying the tab in the field declaration will make it easier to understand and use.

**Comment by tabacitu** [comment]
Merged into 3.2!

3.2 users can now use tabs just by specifying the tab name, inside the field definition. No need to enable the feature, no nothing.

```php
$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 1',
]); // creates tab called "Tab 1"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 2',
]); // creates tab called "Tab 2"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 1',
]); // adds it to "Tab 1"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    'tab' => 'Tab 3',
]); // creates tab called "Tab 3"

$this->crud->addField([
    'label' => 'Option',
    'name' => 'answer_1',
    'type' => 'text',
    'attributes' => ['required' => 'required'],
    // no tab defined
]);  // adds it to "Tab 3"
```

**Comment by MarcosBL** [comment]
Just a minor issue with this: if you add a required email field in a tab, and it doesn't validate, the validation breaks JS as "email" is not a focusable item (you would need to activate the tab first to be able to focus it, as it is hidden).

**Comment by OwenMelbz** [comment]
As far as I can remember there is no JavaScript validation.

I think this might be your browser trying to validate it. Are you adding the "required" attribute to the field?

**Comment by OwenMelbz** [comment]
@marcosBL

If you check the change log https://github.com/Laravel-Backpack/CRUD/pull/289/files

You can see there is no JS for the tabs. So it must be your browser trying to do it

**Comment by MarcosBL** [comment]
Of course you are right @OwenMelbz it was a Chrome extension interfering.

I even tried to set an email field as required, in the second tab, and it worked like a charm:

image

**Pull request merged**